python/triton_kernels/triton_kernels/tensor_details/layout_details/blackwell_scale.py

```diff
@@ -39,55 +39,223 @@ def __init__(self, shape) -> None:
+    vals = tl.load(in_ptrs, mask=in_mask & out_mask, other=0.0)
+
+    # store into output
+    out_ptrs = out_ptr + out_rows * stride_out_m + out_cols * stride_out_n
+    tl.store(out_ptrs, vals, mask=out_mask)
+
+
+def pad_segments_triton(data, ex_hist, ex_hist_padded, K, K_pad, total_padded_rows):
+    """
+    data:           [total_M, K]       (contiguous or with known strides)
+    ex_hist:        [H] int32/long     (segment lengths)
+    ex_hist_padded: [H] int32/long     (padded lengths)
+    returns padded_data: [sum(ex_hist_padded), K_pad]
+    """
+    assert data.is_cuda
+    assert ex_hist.is_cuda and ex_hist_padded.is_cuda
+    assert ex_hist.dim() == 1 and ex_hist_padded.dim() == 1
+    assert ex_hist.shape == ex_hist_padded.shape
+
+    device = data.device
+    H = ex_hist.shape[0]
+
+    ex_hist_int = ex_hist.to(torch.int32)
+    ex_hist_padded_int = ex_hist_padded.to(torch.int32)
+
+    # prefix sums in *rows*
+    # in_offsets[i] = sum_{j < i} ex_hist[j]
+    # out_offsets[i] = sum_{j < i} ex_hist_padded[j]
+
+    # from exp_hist to offset format
+    in_offsets = torch.empty_like(ex_hist_int)
+    out_offsets = torch.empty_like(ex_hist_padded_int)
+
+    if H > 0:
+        in_offsets.zero_()
+        out_offsets.zero_()
+        if H > 1:
+            torch.cumsum(ex_hist_int[:-1], dim=0, out=in_offsets[1:])
+            torch.cumsum(ex_hist_padded_int[:-1], dim=0, out=out_offsets[1:])
+
+    # total_padded_rows = int(ex_hist_padded_int.sum().item())
+    padded_data = torch.empty(total_padded_rows, K_pad, device=data.device, dtype=data.dtype)
+
+    # strides (in elements, not bytes)
+    stride_in_m, stride_in_n = data.stride()
+    stride_out_m, stride_out_n = padded_data.stride()
+
+    BLOCK_M = 32
+    BLOCK_N = 64
+
+    grid = (
+        H,
+        triton.cdiv(128*2, BLOCK_M),
+    )
+
+    pad_segments_kernel[grid](
+        data,
+        ex_hist_int,
+        ex_hist_padded_int,
+        in_offsets,
+        out_offsets,
+        K,
+        K_pad,
+        stride_in_m,
+        stride_in_n,
+        stride_out_m,
+        stride_out_n,
+        padded_data,
+        BLOCK_M=BLOCK_M,
+        BLOCK_N=BLOCK_N,
+    )
+
+    return padded_data
+
 class BlackwellActMXScaleLayout(Layout):
     # Swizzling for activation tensor [M, K], M can be ragged dimension and equals to sum of expert bs
     name: str = "BLACKWELL_SCALE"
 
-    def __init__(self, shape) -> None:
+    def __init__(self, shape, ex_hist: torch.Tensor | None = None) -> None:
         super().__init__(shape)
-        assert len(shape) == 3, f"Only support 3D shape for BlackwellActMXScaleLayout, got {shape}"
-        (
-            *self.leading_shape,
-            self.M,  # sum of expert bs
-            self.K,
-        ) = shape
-        self.B = math.prod(self.leading_shape)
-        self.mode = "batched"
+        if len(shape) == 2:
+            (
+                self.M,  # sum of expert bs
+                self.K,
+            ) = shape
```

Comment:
```suggestion
            self.leading_shape = ()
            (
                self.M,  # sum of expert bs
                self.K,
            ) = shape
```

So that unswizzle works for "ragged"

Reply:
added an assertion to allow only 3D in batched and 2D in ragged (that's the case we care about) so i can just use self.B